x-pack/plugins/ingest_manager/server/services/epm/agent/agent.ts

```diff
@@ -5,12 +5,73 @@
  */
 
 import Handlebars from 'handlebars';
+import { safeLoad } from 'js-yaml';
+import { DatasourceConfigRecord } from '../../../../common';
 
-interface StreamVars {
-  [k: string]: string | string[];
+function isValidKey(key: string) {
+  return key !== '__proto__' && key !== 'constructor' && key !== 'prototype';
 }
 
-export function createStream(vars: StreamVars, streamTemplate: string) {
-  const template = Handlebars.compile(streamTemplate);
-  return template(vars);
+function replaceVariablesInYaml(yamlVariables: { [k: string]: any }, yaml: any) {
+  if (Object.keys(yamlVariables).length === 0 || !yaml) {
+    return yaml;
+  }
+
+  Object.entries(yaml).forEach(([key, value]: [string, any]) => {
+    if (typeof value === 'object') {
+      yaml[key] = replaceVariablesInYaml(yamlVariables, value);
+    }
+
+    if (typeof value === 'string' && value in yamlVariables) {
+      yaml[key] = yamlVariables[value];
+    }
+  });
+
+  return yaml;
+}
+
+function buildTemplateVariables(variables: DatasourceConfigRecord) {
+  const yamlValues: { [k: string]: any } = {};
+  const vars = Object.entries(variables).reduce((acc, [key, recordEntry]) => {
+    // support variables with . like key.patterns
+    const keyParts = key.split('.');
+    const lastKeyPart = keyParts.pop();
+
+    if (!lastKeyPart || !isValidKey(lastKeyPart)) {
+      throw new Error('Invalid key');
+    }
+
+    let varPart = acc;
+    for (const keyPart of keyParts) {
+      if (!isValidKey(keyPart)) {
+        throw new Error('Invalid key');
+      }
+      if (!varPart[keyPart]) {
+        varPart[keyPart] = {};
+      }
+      varPart = varPart[keyPart];
+    }
+
+    if (recordEntry.type && recordEntry.type === 'yaml') {
+      const yamlKeyPlaceholder = `##${key}##`;
+      varPart[lastKeyPart] = `"${yamlKeyPlaceholder}"`;
+      yamlValues[yamlKeyPlaceholder] = recordEntry.value ? safeLoad(recordEntry.value) : null;
+    } else {
+      varPart[lastKeyPart] = recordEntry.value;
+    }
+    return acc;
+  }, {} as { [k: string]: any });
+
+  return { vars, yamlValues };
+}
+
+export function createStream(variables: DatasourceConfigRecord, streamTemplate: string) {
+  const { vars, yamlValues } = buildTemplateVariables(variables);
+
+  const template = Handlebars.compile(streamTemplate, { noEscape: true });
```

Comment:
Sorry, I thought this was also updated. Mind keeping the escaping?
```suggestion
  const template = Handlebars.compile(streamTemplate);
```